### PR TITLE
Improve LaTeX footnotes

### PR DESCRIPTION
**Pull request by jfbu**
The footnotes in PDF output are improved in the following ways:
1. they are allowed to contain literal-blocks, even with a caption ! (and cross references to that caption in the footnote will work)
2. they more often give working clickable hyperlinks,
3. the footnote marker for a footnote appearing in a section heading does not contaminate the tables of contents nor the PDF bookmarks.

It is a limitation of hyperref that the `\footnotemark[N]/\footnotetext[N]` syntax gives non-hyperlinked footnotes. Currently this syntax is used in the environments for table material (threeparttable with tabular or tabulary, longtable). But it is only needed in the captions and in the cells of tabulary. But I don't know how to modify latex.py to handle more liberally the footnotes because the code in latex.py decides only on leaving the table if it will be rendered by tabulary or by longtable. Thus currently footnotes from inside a longtable will use the `\footnotemark[N]/\footnotetext[N]` and will not be hyperlinked, although only the table heads need this restriction.

As per the rendering of literal-blocks inside footnotes, I have maintained the framing with some hesitation about the design (it is a one line change to modify that.) And the caption is still printed in a centered position (it would require more changes to sphinx.sty to customize that currently.)

With this PR, Sphinx stops shipping a modified tabulary LaTeX package. A footnotehyper-sphinx.sty package replaces it: it corrects bugs from package footnote.sty which has been used by Sphinx for a long time, in particular with respect to hyperref and the possibility to use the environment form of footnote.

### CODE REVIEW

**Comment by jfbu** [comment]
The latest commit opts for using upstream footnotehyper with no modification. Maintenance will be easier.

File: sphinx/texinputs/sphinx.sty [hunk 1]
```diff
@@ -47,11 +47,24 @@
...
+    {\ifmeasuring@\expandafter\@gobbletwo\else\expandafter\@firstofone\fi}%
+    {\ifx\equation$\expandafter\@gobbletwo\fi\FNH@hyper@fntext@i }%$
+   }%
+% allow footnotes in titles (silently removed from TOCs and PDF bookmarks)
+   \def\sphinxfootnotemark [#1]%
+     {\ifx\thepage\relax\else \protect\footnotemark[#1]\fi}%
+% ensure hyperref compatibility
+   \AtBeginDocument
+       {\@ifpackageloaded{hyperref}
+       {\pdfstringdefDisableCommands{\def\sphinxfootnotemark [#1]{}}}{}}%
```
Comment (jfbu) [inline sphinx/texinputs/sphinx.sty]: actually, use of hyperref is deeply embedded inside Sphinx, hence this check for hyperref in use is a  bit rhetorical.

File: sphinx/writers/latex.py [hunk 2]
```diff
@@ -1873,9 +1873,14 @@ def warner(msg):
                                                       **highlight_args)
             # workaround for Unicode issue
             hlcode = hlcode.replace(u'€', u'@texteuro[]')
+            if self.in_footnote:
+                self.body.append('\n\\sphinxSetupCodeBlockInFootnote')
+                hlcode = hlcode.replace('\\begin{Verbatim}',
+                                        '\\begin{sphinxVerbatim}')
+                pass
```
Comment (tk0miya) [inline sphinx/writers/latex.py]: Please remove this line.

File: sphinx/writers/latex.py [hunk 3]
```diff
@@ -1886,9 +1891,10 @@ def warner(msg):
             # get consistent trailer
             hlcode = hlcode.rstrip()[:-14]  # strip \end{Verbatim}
             self.body.append('\n' + hlcode + '\\end{sphinxVerbatim%s}\n' %
-                             (self.table and 'intable' or ''))
+                             ((self.table and not self.in_footnote) and
```
Comment (tk0miya) [inline sphinx/writers/latex.py]: I feel this condition become complex. It's a bit unreadable. So could you rewrite this to simple if-else notation?

**Review by tk0miya — COMMENTED** [review]
That sounds great!
LGTM with nits.

**Comment by jfbu** [comment]
@tk0miya thank you for review. Sphinx already used "foonote" package, but not its environment footnote which has the potential to allow verbatim material. The "footnotehyper-sphinx" which I am contributing fixes the bugs in "footnote" (which made its environment unusable) and adds hyperref compatibility. I hope the switch to using `footnote/footnotetext+\footnotemark` environment rather than `\footnote/\footnotetext+\footnotemark` commands will not create issues for people. It opens up possibility for code-blocks inside footnotes, and makes (except in tables) footnotes hyperlinked.

I am merging now for 1.5a2 so that the feature can be tested in time by developers. I squash all commits into one for the merge for cleaner history.

**Pull request merged**